server/mqtt.go

```diff
@@ -4271,32 +4297,32 @@ func mqttDeliverMsgCbQoS0(sub *subscription, pc *client, _ *Account, subject, re
 		return
 	}
 
-	hdr, msg := pc.msgParts(rmsg)
-
 	// This is the client associated with the subscription.
 	cc := sub.client
 
 	// This is immutable
 	sess := cc.mqtt.sess
 
-	// Check the subscription's QoS. This needs to be protected because
-	// the client may change an existing subscription at any time.
-	sess.mu.Lock()
+	// Lock here, otherwise we may be called with sub.mqtt == nil. Ignore
+	// wildcard subscriptions if this subject starts with '$', per Spec
+	// [MQTT-4.7.2-1].
+	sess.subsMu.RLock()
```

Comment:
The 2nd part of "the fix", replacing sess.mu with sess.subsMu here. We need to ensure that `sub.mqtt` has been set up, not a nil; and we need the QOS.

Reply:
But it would not prevent the callback to proceed even if retained messages are not fully serialized. When calling `mqttEnqueuePublishMsgTo()` at the end of that function, there will be a check to see if the `prm` writer has something and then write it out before the messages. But that relied on the fact that they would be done under the session lock to prevent this "race" of the callback firing before we are ready.

Reply:
FWIW, I still think that the dedicated granular RWMutex for QOS0 makes sense since it's the "hot" callback; but yes retained messages need to be queued for delivery first, mosquitto does it, I checked. I'll add a test for this, it's easy on a cluster since there are most definitely many "internal" NATS messages in flight while processing an MQTT subscription.

@derekcollison's thought was that maybe the MQTT code (the QOS0 callback, what else?) is slowing down the route message processing. I will re-synchronize the writing out of the retained messages and benchmark it.

0/5 not sure if it's worth it, but maybe make mqttWriter a [][]byte with a `queueToClientOutbound()` that would queue up chunk by chunk (it all gets re-buffered again there anyway?). Or would it be too expensive to call `c.queueOutbound` for small chunks? Ditto in https://github.com/nats-io/nats-server/blob/e8772d592cab4d503fe9744ca73e5af2ffd155d0/server/mqtt.go#L4466, is there a reason for re-buffering the message and calling `c.queueOutbound` once, rather than separately for the header, and then for  msg itself? We may be copying a lot of bytes there, with a `#` sub.

Another 0/5. I understand that we consider NATS->MQTT QOS0 to be an essential feature, but should we put some constraints on it? I have tried naive blacklisting attempts, maybe it's a direction worth exploring more? Are there more kinds of (cluster-specific?) internal messages that we should "firewall out" of the MQTT code altogether?

Reply:
You mean start to queueOutbound retained messages one by one instead of using the `prm`? That we can't because we want the retained messages to go before incoming messages, but they have to go after the SUBACK. This is why I used this buffer. 
As to make the writer a [][]byte instead of a single writer, maybe, but not sure how this is going to change much. But you can experiment and see... Also, to see any difference my guess is that you would need to get a LOT of retained messages (that is, lot of different topics, etc..).

Reply:
> You mean start to queueOutbound retained messages one by one instead of using the prm? That we can't because we want the retained messages to go before incoming messages, but they have to go after the SUBACK. This is why I used this buffer.

No, that's understood, and understood well now. I was just suggesting that I keep the new `subsMu` lock specifically for QoS0 delivery, and move the rms->prm serialization under it, potentially with the [][] optimization so we don't need to copy the bytes while holding the lock. 

> you would need to get a LOT of retained messages (that is, lot of different topics, etc..).

Agreed. But lots of messages can happen with wildcard subscriptions, and their size can be considerable. Hard to estimate without the use-case (and IMO it'd be a sub-optimally architected app to do this kind of thing with large messages, but I don't know...)

Reply:
If serialization is under subsMu, then it could work, but then I am not sure that will minimize by that much, but only experimentation will tell.

Reply:
I changed it now so that `subsMu` covers just the initialization of the (mqtt) subs, and there is a per-sub WaitGroup indicating that its `prm` is ready.

Reply:
I am going to resolve this thread since there have been a lot of changes since, and the core issue (serializing and sending prm to a sub before other messages) should have been addressed.